src/branch_and_bound.rs

```diff
@@ -0,0 +1,582 @@
+use bitcoin::Amount;
+
+use crate::WeightedUtxo;
+use bitcoin::amount::CheckedSum;
+use bitcoin::blockdata::effective_value;
+use bitcoin::FeeRate;
+
+fn index_to_utxo_list(
+    index_list: Option<Vec<usize>>,
+    weighted_utxos: &mut [WeightedUtxo],
+) -> Option<Vec<WeightedUtxo>> {
+    index_list.map(|i_list| i_list.iter().map(|i: &usize| weighted_utxos[*i].clone()).collect())
```

Comment:
I realized this collects the vec which may be unneeded for the callers, so it's better tor return `impl Iterator<Item=WeightedUtxo>` or even better `impl Iterator<Item=&WeightedUtxo>` if we can get the borrow checker to agree.

Reply:
What about:

```rust
fn index_to_utxo_list<'a>(
    index_list: Option<Vec<usize>>,
    weighted_utxos: &'a mut [&'a WeightedUtxo],
) -> Option<impl Iterator<Item=&'a WeightedUtxo>> {
    index_list.map(move |i_list| i_list.into_iter().map(move |i: usize| weighted_utxos[i]))
}
```

and the corresponding test
```rust
    #[test]     
    fn index_to_utxo_list_collect() {
        let mut wu = [&WeightedUtxo {
            satisfaction_weight: Weight::ZERO,
            utxo: TxOut { value: Amount::from_str("1 sat").unwrap(), script_pubkey: ScriptBuf::new() }
        }];
        let i = Some(vec![0]);
        let list: Vec<_> = index_to_utxo_list(i, &mut wu).unwrap().collect();
    }   
```

I think `fn select_coins_bnb(` should use the same return type then `Option<impl Iterator<Item=&'a WeightedUtxo>>`

Reply:
Yeah, looks good though I think `&mut [&'a WeightedUtxo]` should be used instead (notice missing lifetime). There shouldn't be the need to force the outer reference into the lifetime, only the inner ones. (But this reaches the limits of my understanding of lifetimes so maybe the borrow checker disagrees.)

Reply:
@Kixunil I ended up changing the return type to:
) -> Option<std::vec::IntoIter<&'_ WeightedUtxo>> {

here: https://github.com/p2pderivatives/rust-bitcoin-coin-selection/pull/30

The opaque type:  `impl Iterator<Item=&'a WeightedUtxo>` was giving me problems in `lib.rs`.  The compiler won't allow both SRD and BNB to return opaque types.